### PR TITLE
Feature/road segments

### LINKED ISSUE #35: Process to network graph representation for roads
We import OSM data for highways, which comes defined as a set of nodes and a set of ways. A way's geometry is defined by a list of nodes and can be interpreted as a LineString. For routing over the highway network, we need to interpret this as a network graph, which can be defined as a set of edges corresponding to segments of highway (either all or part of a way) and nodes corresponding to highway junctions and endpoints (some of the OSM nodes are nodes in this sense, others just go into defining the edge segment geometry).

Aside: There are other representations specialised for fast response to A-B routing queries which provide a series of directions (turn left at the end of the road, third exit on the roundabout), particularly the [edge-expanded graph representation](https://github.com/Project-OSRM/osrm-backend/wiki/Graph-representation) used in OSRM. I don't think we want these, as our routing queries will mostly be bulk source-to-multiple-target in the context of transport simulation modelling, and the physical edges are the things that would be disrupted by flooding or other hazards.

Detail note: When creating box extracts from OSM data, the osmium tool provides a couple of ways of dealing with ways which extend beyond the boundaries of the box (https://osmcode.org/osmium-tool/manual.html#extract-strategies) and defaults to "complete ways", which includes all of a long road section in an extract if any part of it is within the box. We should handle roads which cross a box boundary by adding a node on the boundary, and leaving the part outside of the box to be handled inside of the box it belongs to.

Done looks like: each OSM highway extract box can be processed into a graph representation, stored in two parquet files: nodes and edges, where each maintains the source OSM ID and attributes, but nodes only includes junctions and endpoints, and edges may have been split into multiple segments at junctions.

### PR DESCRIPTION
**Pull request by mjaquiery**
Closes #35

### CODE REVIEW

File: workflow/scripts/intersection.py [hunk 1]
```diff
@@ -157,26 +155,25 @@ def write_empty_files(columns, outputs_path):
...
-    attrs = attrs.split(",")
+        # print(sys.argv)
+        # (
+        #     network_edges_path,
+        #     hazard_dir,
+        #     output_path,
+        # ) = sys.argv[1:]
+        network_edges_path = '../../results/geoparquet/tanzania-mini_filter-highway-core/slice-2.geoparquet'
+        output_path = '../../results/test.geoparquet'
+        hazard_dir = '../../results/input/hazard-aqueduct-river/tanzania-mini'
```
Comment (tomalrussell) [inline workflow/scripts/intersection.py:169]: Is this a testing/debugging change that has crept in?

Comment (mjaquiery) [reply workflow/scripts/intersection.py:169]: Good spot!

Comment (mjaquiery) [reply workflow/scripts/intersection.py]: This is now fixed and tests have been updated. The intersection tests takes ~45s to run, nearly half the total test time -- I can't remember if it always did that, but adding extra info to geoparquet input files shouldn't cause that process to take longer than before.

**Comment by mjaquiery** [comment]
This should now also close #53 with the recent pushes.

**Review by tomalrussell — APPROVED** [review]
Looks good to me - the "way parsing to network" structure in `osm_to_pq.py` makes sense, I'll be keen to try this out in a few places and explore how the network structure comes out as parsed directly from OSM.

File: workflow/scripts/join_data.py [hunk 2]
```diff
@@ -33,6 +35,62 @@ def append_data(base, slice_files):
...
+
+        return {
+            'changed': changed,
+            'base': base
+        }
+
+    node_ref = -1
+    max_cycles = 100000000
+    while max_cycles > 0:
+        max_cycles += 1
```
[the remark this answers is not in the document]
Comment (mjaquiery) [reply workflow/scripts/join_data.py]: Good spot!

**Pull request merged**